### PR TITLE
Custom Navigation Bar

### LINKED ISSUE #460: Add tab bar
**Description:**
- Add tab bar

### PR DESCRIPTION
**Pull request by jackthmp**
Closes #460, #455, #452 

- Subclassed ViewController with navigation bar types: `default`, `custom`, `hidden`
- Subclassed NavigationBar to set up upon init
- Added tab bar
- Created topic cells, no `Topic` data structure yet
- Added some colors and patterns
- Reorganized file structure

### CODE REVIEW

File: Navigation Bars/CustomNavigationBar.swift [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    /// Otherwise bar is a solid color.
+    var isTranslucent: Bool = true {
+        didSet {
+            backgroundView.alpha = isTranslucent ? 0.6 : 1.0
+            blurEffectView.effect = isTranslucent ? blurEffect : nil
+        }
+    }
+
+    /// Color of the navigation bar.
+    var barColor: UIColor = .black {
```
Comment (mindylou) [inline Navigation Bars/CustomNavigationBar.swift]: wow this is next level documentation

File: Recast/Controllers/TabBarController.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        discoverNavController = NavigationController(rootViewController: discoverViewController)
+
+        let navControllers: [NavigationController] = [homeNavController, discoverNavController]
+        setViewControllers(navControllers, animated: true)
+        previousViewController = navControllers[selectedIndex].viewControllers.first
+    }
+
+    /// Scrolls to top upon tap of current tab.
+    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
+//        if let navigationController = viewController as? UINavigationController,
```
Comment (mindylou) [inline Recast/Controllers/TabBarController.swift]: delete this stuff if it's unused rn

File: Recast/Utilities/ViewController.swift [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+//  Recast
+//
+//  Created by Jack Thompson on 11/20/18.
+//  Copyright © 2018 Cornell AppDev. All rights reserved.
+//
+
+import UIKit
+
+enum NavBarType {
+    case `default`
```
Comment (mindylou) [inline Recast/Utilities/ViewController.swift:12]: what exactly do all of these types do?

File: Recast/Utilities/ViewController.swift [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            if navBarType == .custom {
+                mainScrollView?.delegate = self
+                mainScrollView?.contentInsetAdjustmentBehavior = .never
+                mainScrollView?.contentInset = self.additionalSafeAreaInsets
+            }
+        }
+    }
+
+//    /// Guarantees that navigation bar is set upon load of view to avoid lag.
+//    override func willMove(toParent parent: UIViewController?) {
```
Comment (mindylou) [inline Recast/Utilities/ViewController.swift]: delete if unnecessary

File: Recast/Utilities/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        var newSafeArea = view.safeAreaInsets
+        newSafeArea.top = CustomNavigationBar.height
+        newSafeArea.bottom = AppDelegate.appDelegate.tabBarController.tabBar.frame.height
+        self.additionalSafeAreaInsets = newSafeArea
+        viewSafeAreaInsetsDidChange()
+    }
+
+    /// Resizes Custom Navigation Bar upon scrolling, if `navBarType = .custom`.
+    dynamic func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        customNavBar?.snp.updateConstraints({ update in
```
Comment (mindylou) [inline Recast/Utilities/ViewController.swift]: i think it can just be `updateConstraints { update in `, no parentheses

**Review by mindylou — COMMENTED** [review]
after these are fixed LGTM!

**Review by mindylou — DISMISSED** [review]

**Review by doctor3w — DISMISSED** [review]

**Pull request merged**